### PR TITLE
Returning an empty tensor of param dtype for wgrad

### PR DESCRIPTION
**Pull request by sanandaraj5597**
For the M-LM usage of TE API's, custom layer Backward implementations return the accumulated gradient which is of weight.main_grad dtype. This makes Torch insert an unwanted at::to kernel, this is unwanted as wgrad is unused and is already added to the main grad by the TE API. To remove this redundancy, we return an empty tensor of the shape of weight.main_grad shape and weight.dtype. The empty tensor is later freed in the M-LM code.
Have edited layernorm_linear.py, linear.py and layernorm_mlp.py

### CODE REVIEW

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 1]
```diff
@@ -525,6 +525,11 @@ def backward(
             # Handle custom DDP from mcore.
             if ctx.fuse_wgrad_accumulation and hasattr(weight, 'grad_added_to_main_grad'):
                 weight.grad_added_to_main_grad = True
+                wgrad = torch.empty(weight.main_grad.shape,
+                                   dtype=weight.dtype,
+                                   device=torch.cuda.current_device(),
+                                   requires_grad=False
+                                   )
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/layernorm_linear.py:532]: Could we set `wgrad` to `None`? It seems that this would avoid some memory allocation and would make it clear that the grad isn't supposed to be used.

Comment (sanandaraj5597) [reply transformer_engine/pytorch/module/layernorm_linear.py:532]: Setting wgrad to None, would break the M-LM usage. M-LM has a check that wgrad should not be None. (https://github.com/NVIDIA/Megatron-LM/blob/5f2877d85cb26e47ce6dcdae4b80adf376abf4e8/megatron/core/distributed/distributed_data_parallel.py#L175)
It is anyway freed after this check, this is done due to some multi-threading reason done my Torch during Bprop.

Comment (deepakn94) [reply transformer_engine/pytorch/module/layernorm_linear.py:532]: No, we can't set `wgrad` to `None`...we are trying to prevent PyTorch from launching backward hooks in helper threads (we use backward hooks to tell us when we can launch asynchronous all-reduces / reduce-scatters).

**Review by ksivaman — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/module/linear.py [hunk 2]
```diff
@@ -465,6 +465,11 @@ def backward(
             # Handle custom DDP from mcore.
             if ctx.fuse_wgrad_accumulation and hasattr(weight, 'grad_added_to_main_grad'):
                 weight.grad_added_to_main_grad = True
+                wgrad = torch.empty(weight.main_grad.shape,
+                                   dtype=weight.dtype,
```
Comment (deepakn94) [inline transformer_engine/pytorch/module/linear.py:469]: Nit: indentation off by 1 space.

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 3]
```diff
@@ -525,6 +525,11 @@ def backward(
             # Handle custom DDP from mcore.
             if ctx.fuse_wgrad_accumulation and hasattr(weight, 'grad_added_to_main_grad'):
                 weight.grad_added_to_main_grad = True
+                wgrad = torch.empty(weight.main_grad.shape,
+                                   dtype=weight.dtype,
```
Comment (deepakn94) [inline transformer_engine/pytorch/module/layernorm_linear.py:529]: Nit: indentation off by 1 space.

**Comment by ptrendx** [comment]
How will that affect non MLM codebases?

**Comment by deepakn94** [comment]
> How will that affect non MLM codebases?

Shouldn't affect them (i.e., we won't even return an empty tensor) because the `grad_added_to_main_grad` attribute will not be set.

**Comment by sanandaraj5597** [comment]
Are we good to merge?

File: transformer_engine/pytorch/module/layernorm_mlp.py [hunk 4]
```diff
@@ -888,6 +893,12 @@ def backward(
             # Handle custom DDP from mcore.
             if ctx.fuse_wgrad_accumulation and hasattr(fc2_weight, 'grad_added_to_main_grad'):
                 fc2_weight.grad_added_to_main_grad = True
+                fc2_wgrad = torch.empty(fc2_weight.main_grad.shape,
+                                        dtype=fc2_weight.dtype,
+                                        device=torch.cuda.current_device(),
+                                        requires_grad=False
+                                        )
+
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/layernorm_mlp.py:901]:
```suggestion

```
